lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -26,54 +26,87 @@ namespace firrtl {
+                    << SymbolTable::getSymbolName(op) << "\n");
   auto addSym = [&](StringAttr name, InnerSymTarget target) {
     LLVM_DEBUG(dbgs() << " - @" << name << " -> " << target << "\n");
     assert(name && !name.getValue().empty());
-    auto it = symbolTable.insert({name, target});
-    if (!it.second) {
-      auto orig = symbolTable.lookup(name);
-      (target.getOp()->emitError("duplicate symbol @") << name << " found")
-              .attachNote(orig.getOp()->getLoc())
-          << " symbol also found within this operation";
-      // TODO: rework so can indicate failure to caller, for use in verif/etc?
-    }
-    assert(it.second && "repeated symbol found");
+    return callback(name, target);
   };
-  auto addSyms = [&](InnerSymAttr symAttr, InnerSymTarget baseTarget) {
+
+  auto addSyms = [&](InnerSymAttr symAttr,
+                     InnerSymTarget baseTarget) -> LogicalResult {
     if (!symAttr)
-      return;
+      return success();
     assert(baseTarget.getField() == 0);
     for (const auto &symProp : symAttr.getProps()) {
-      addSym(symProp.getName(), InnerSymTarget::getTargetForSubfield(
-                                    baseTarget, symProp.getFieldID()));
+      if (failed(
+              addSym(symProp.getName(), InnerSymTarget::getTargetForSubfield(
+                                            baseTarget, symProp.getFieldID()))))
+        return failure();
     }
+    return success();
   };
 
   // Walk the operation and add InnerSymbolTarget's to the table.
-  op->walk([&](Operation *curOp) {
-    if (auto symOp = dyn_cast<InnerSymbolOpInterface>(curOp))
-      addSyms(symOp.getInnerSymAttr(), InnerSymTarget(symOp));
-
-    // Check for ports
-    // TODO: investigate why/confirm ports having empty-string symbols is normal
-    // TODO: Add fields per port, once they work that way (use addSyms)
-    if (auto mod = dyn_cast<FModuleLike>(curOp)) {
-      for (const auto &p : llvm::enumerate(mod.getPorts()))
-        if (auto sym = p.value().sym; sym && !sym.getValue().empty())
-          addSym(p.value().sym, InnerSymTarget(p.index(), curOp));
-    }
-  });
+  return success(
+      !op->walk<mlir::WalkOrder::PreOrder>([&](Operation *curOp) -> WalkResult {
+           if (auto symOp = dyn_cast<InnerSymbolOpInterface>(curOp))
+             if (failed(
+                     addSyms(symOp.getInnerSymAttr(), InnerSymTarget(symOp))))
+               return WalkResult::interrupt();
+
+           // Check for ports
+           // TODO: Add fields per port, once they work that way (use addSyms)
+           if (auto mod = dyn_cast<FModuleLike>(curOp)) {
+             for (const auto &p : llvm::enumerate(mod.getPorts()))
+               if (auto sym = p.value().sym; sym && !sym.getValue().empty())
+                 if (failed(addSym(p.value().sym,
+                                   InnerSymTarget(p.index(), curOp))))
+                   return WalkResult::interrupt();
```

Comment:
Hmm, not sure I follow, mostly because still need to know what port (index) the symbol is for (and because apparently that returns an `ArrayRef<Attribute>` not `ArrayRef<StringAttr>` (or `ArrayRef<InnerSymAttr>`)).  Here's the sort of thing I ended up w/`getPortSymbols`:

```
for (const auto &p : llvm::enumerate(mod.getPortSymbols()))
  if (auto sym = p.value().cast<StringAttr>();
      sym && !sym.getValue().empty() /* quirk of port symbols */)
if (failed(walkSym(sym, InnerSymTarget(p.index(), curOp))))
  return WalkResult::interrupt();
```

Maybe this means we need better way to get port symbols.. :+1: .

Reply:
Yeah I was mostly thinking that `getPorts` seems kind of expensive, since it constructs a small vector with lots of different information about each port.  Did you try the similar `getPortSymbolsAttr().getAsRange<StringAttr>()`? (does that still work with enums?)

Another idea is to use a regular for loop:
```
for (int i = 0, e = numPorts; i < e) {
  if (failed(addSym(mod.getPortSymbolAttr(i);
                                   InnerSymTarget(i, curOp))))
                   return WalkResult::interrupt();
}
```

Reply:
Thanks!